haskell/toolchain.bzl

```diff
@@ -56,6 +70,41 @@ def _run_ghc(hs, cc, inputs, outputs, mnemonic, arguments, env, params_file = No
     hs.actions.write(compile_flags_file, args)
     hs.actions.write(extra_args_file, arguments)
 
+    if abi_file != None:
+        # We declare the file containing informations about the call to GHC with --show-iface.
+        show_iface_file = hs.actions.declare_file("show_iface_%s_%s" % (hs.name, extra_name))
+        show_iface_args = hs.actions.args()
+        show_iface_args.set_param_file_format("multiline")
+        show_iface_args.add(hs.tools.ghc.path)
+        show_iface_args.add("--show-iface ")
+        show_iface_args.add(hi_file.path)
+        hs.actions.write(show_iface_file, show_iface_args)
+
+        # We create a file containing the name of all the interface files which should not be considered
+        # by the caching mechanism to know if recompilation should be triggered.
+        # This behaivour is extensively described in the Note [On the ABI hash] in haskell/experimaental/private/module.bzl
+        interface_with_abis_list = hs.actions.declare_file("interfaces_%s_%s" % (hs.name, extra_name))
+        interface_with_abis_args = hs.actions.args()
+        interface_with_abis_args.set_param_file_format("multiline")
+        interface_with_abis_args.add_all(interface_inputs)
+        hs.actions.write(interface_with_abis_list, interface_with_abis_args)
+
+        # The interface_with_abis_list should be used as unused_inputs_list.
+        # However, since it is more traditional to generate the unused_inputs_list within the action itself,
+        # and this proposition is already quite hackish, we wanted to stick to the usage,
+        # expecting it to offer us a stronger reliability, even with the future evolution of Bazel.
+        unused_inputs_list = hs.actions.declare_file("unused_%s_%s" % (hs.name, extra_name))
```

Comment:
How about making the new declared files that are introduced in this PR use paths that depend on all the parameters that the earlier files did? For instance, all of the following depend on `hs.name` and `extra_name`
```
show_iface_file = hs.actions.declare_file("show_iface_%s_%s" % (hs.name, extra_name))
interface_with_abis_list = hs.actions.declare_file("interfaces_%s_%s" % (hs.name, extra_name))
unused_inputs_list = hs.actions.declare_file("unused_%s_%s" % (hs.name, extra_name))
```
while the preexisting files depend on `hs.name`, `extra_name`, and `mnemonic`.
```
compile_flags_file = hs.actions.declare_file("compile_flags_%s_%s_%s" % (hs.name, extra_name, mnemonic))
extra_args_file = hs.actions.declare_file("extra_args_%s_%s_%s" % (hs.name, extra_name, mnemonic))
```

Reply:
Turns out this is not necessary as we are only creating this files with a single mnemonic so far.